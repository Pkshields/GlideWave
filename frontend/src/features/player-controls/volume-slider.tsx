import { useState } from "react"
import { repeat } from "../../utils/loops"

export function VolumeSlider() {
    const [volume, setVolume] = useState(10)
    const [hoverOverNotch, setHoverOverNotch] = useState(-1)

    function generateSliderNotchClasses(index: number): string {
        const styleClassNames = "py-3 px-1 mx-0.5 border-2 rounded border-white scale-75"

        const effectiveVolume = hoverOverNotch == -1 ? volume : hoverOverNotch
        if (index <= effectiveVolume - 1) {
            return `${styleClassNames} bg-white`
        }

        return styleClassNames
    }

    return (
        <div className="flex space-x-0 items-center">
            {repeat(
                index => (
                    <div
                        data-testid={`volume-notch-${index}`}
                        className="cursor-pointer"
                        onMouseEnter={() => setHoverOverNotch(index + 1)}
                        onMouseLeave={() => setHoverOverNotch(-1)}
                        onClick={() => setVolume(index + 1)}
                    >
                        <div className={generateSliderNotchClasses(index)} />
                    </div>
                ),
                10
            )}
        </div>
    )
}