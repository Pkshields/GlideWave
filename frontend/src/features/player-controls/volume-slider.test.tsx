import { afterEach, describe, expect, it } from "vitest"
import { VolumeSlider } from "./volume-slider"
import { cleanup, render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

describe("volume slider ui tests", () => {
    afterEach(cleanup)

    it("should start with all notches highlighted", () => {
        const component = render(<VolumeSlider />)

        expect(component.container).toMatchSnapshot()
    })

    it("should show current volume level after volume is changed", async () => {
        const component = render(<VolumeSlider />)

        await userEvent.click(screen.getByTestId('volume-notch-5'))

        expect(component.container).toMatchSnapshot()
    })

    it("should show potential volume if notch is highlighted", async () => {
        const component = render(<VolumeSlider />)

        await userEvent.hover(screen.getByTestId('volume-notch-6'))

        expect(component.container).toMatchSnapshot()
    })

    it("should show current volume when notch is unhighlighted", async () => {
        const component = render(<VolumeSlider />)

        await userEvent.hover(screen.getByTestId('volume-notch-3'))
        await userEvent.unhover(screen.getByTestId('volume-notch-3'))

        expect(component.container).toMatchSnapshot()
    })
})