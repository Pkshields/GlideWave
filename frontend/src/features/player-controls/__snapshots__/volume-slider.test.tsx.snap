// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`volume slider ui tests > should show current volume level after volume is changed 1`] = `
<div>
  <div
    class="flex space-x-0 items-center"
  >
    <div
      data-testid="volume-notch-0"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-1"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-2"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-3"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-4"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-5"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-6"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75"
      />
    </div>
    <div
      data-testid="volume-notch-7"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75"
      />
    </div>
    <div
      data-testid="volume-notch-8"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75"
      />
    </div>
    <div
      data-testid="volume-notch-9"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75"
      />
    </div>
  </div>
</div>
`;

exports[`volume slider ui tests > should show current volume when notch is unhighlighted 1`] = `
<div>
  <div
    class="flex space-x-0 items-center"
  >
    <div
      data-testid="volume-notch-0"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-1"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-2"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-3"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-4"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-5"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-6"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-7"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-8"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-9"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
  </div>
</div>
`;

exports[`volume slider ui tests > should show potential volume if notch is highlighted 1`] = `
<div>
  <div
    class="flex space-x-0 items-center"
  >
    <div
      data-testid="volume-notch-0"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-1"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-2"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-3"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-4"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-5"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-6"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-7"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75"
      />
    </div>
    <div
      data-testid="volume-notch-8"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75"
      />
    </div>
    <div
      data-testid="volume-notch-9"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75"
      />
    </div>
  </div>
</div>
`;

exports[`volume slider ui tests > should start with all notches highlighted 1`] = `
<div>
  <div
    class="flex space-x-0 items-center"
  >
    <div
      data-testid="volume-notch-0"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-1"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-2"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-3"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-4"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-5"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-6"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-7"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-8"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
    <div
      data-testid="volume-notch-9"
    >
      <div
        class="py-3 px-1 mx-0.5 border-2 rounded border-white scale-75 bg-white"
      />
    </div>
  </div>
</div>
`;
